### DISCUSSION TITLE
inconsistent startup behavior on ESP32 localtime

### DISCUSSION THREAD

**Discussion by wohltat** [micropython/micropython#13217] [ESP32]
I see different boot behavior depending on if i soft reboot (Ctrl + D) or power cycle (plug in)
Soft reboot:
- time.localtime() is automatically synchronized to my local time, even without `ntptime.settime()`

Power cycle:
- time.localtime() is not automatically synchronized and time begins on 2000-01-01 00:00:00
- i can use ntptime.settime() to synchronize the time. But it is different time to soft reboot. Seems to use a different time zone (UTC vs my timezone).

Questions:
- Why is there a different behavior?
- Where does the time come from in the soft reboot case? Should'nt the time be the same?
- Is there a way to set the offset for `time.localtime()`? Seems i have to do it manually. 

The Code i used:
``` python
# boot.py
import network
import settings
import time
import ntptime
from errno import EHOSTUNREACH

def log(s = None):
    if s is None:
        s = ''
    with open('log.txt', 'a') as f:
        current_time = time.localtime()
        
        # Format the date and time without strftime
        formatted_date = "{:04d}-{:02d}-{:02d} {:02d}:{:02d}:{:02d}".format(
            current_time[0], current_time[1], current_time[2],
            current_time[3], current_time[4], current_time[5]
        )
        f.write(formatted_date + ', ' + str(s) + '\n')
        
log('_______________')
sta_if = network.WLAN(network.STA_IF);
sta_if.active(True)

print('connect')
sta_if.connect(settings.WIFI_SSID, settings.WIFI_PASSWORD)

while not sta_if.isconnected():
    time.sleep(1)
    log('.')
print('connected')
log(sta_if.ifconfig())

while True:
    try:
        ntptime.settime()
        print('ntp sync')
        break
    except OSError as e:
        if e.errno == EHOSTUNREACH:
            print('EHOSTUNREACH')
            time.sleep(5)

print(time.localtime())
log(sta_if.ifconfig())


```

---------
```
MicroPython v1.22.0-preview.73.g7736ab832.dirty on 2023-11-24; Generic ESP32 module with ESP32
Type "help()" for more information.
```

**Comment by davefes** [comment]
For soft-reboot and if you are connected using rshell the local time is obtained from the host. 

I use to use a program to change ntp time into local time, but as you might want to be able to account for daylight saving have a look for:
`daylight saving micropython`

https://gist.github.com/Beormund/f0f39c72e066da497f6308d1964c9627

**Reply by glenn20** [reply]
Also, soft reset does not reset the RTC, so it will continue to keep time as it had been configured before the soft reset. A soft reset only restarts micropython, it does not reset the hardware. From the esp32 hardware and cpu perspective, the application continues to run normally.

**Reply by bulletmark** [reply]
@davefes AFAIK mpremote never sets the device time, unless you explicitly issue a `rtc --set` command. Not sure about rshell but I know that thonny does sync the time from your host pc.

**Reply by wohltat** [reply]
That explains it. I forgot to say that I was using thonny. So I guess it's a bug in thonny then, since it is using a different time, i.e. not UTC?

**Reply by davefes** [reply]
@bulletmark, Thanks for the correction. rshell does sync time from the host.
